### PR TITLE
Task 104: Implement 'scope-up' and 'scope-down' CLI Commands for Dynamic Task Complexity Adjustment

### PR DESCRIPTION
**Pull request by eyaltoledano**
## Task Overview
Implementation of scope-up and scope-down CLI commands for dynamic task complexity adjustment with strength levels, custom prompts, and MCP integration.

## Subtasks Completed
- [x] 104.1: Extended CLI with scope-up and scope-down commands
- [x] 104.2: Implemented task/subtask retrieval and context gathering  
- [x] 104.3: Developed core complexity adjustment logic
- [x] 104.4: Implemented task update and persistence logic
- [x] 104.5: Integrated with MCP tool and added comprehensive error handling

## Implementation Details
- Full CLI command integration with strength parameters (light/regular/heavy)
- MCP tool equivalents for integrated environments like Cursor
- Comprehensive test coverage using TDD methodology
- AI-powered complexity adjustment with context awareness
- Robust error handling and validation
- Added new CLI commands: `task-master scope-up` and `task-master scope-down`
- Support for comma-separated task/subtask IDs
- Custom prompt functionality for targeted adjustments
- Three strength levels: light, regular, heavy

## Testing
- [x] Unit tests added with full TDD coverage
- [x] Integration tests for CLI and MCP interfaces  
- [x] Manual testing completed across different scenarios
- [x] All existing tests still passing

## Breaking Changes
None - purely additive functionality.

## Related Tasks
Completes Task 104 and enables dynamic task complexity management workflows.

## Files Added/Modified
- **New MCP Tools**: scope-up.js, scope-down.js  
- **New Core Logic**: scope-adjustment.js
- **CLI Integration**: Updated commands.js and task-manager.js
- **Test Coverage**: scope-adjustment.test.js
- **Documentation**: test_workflow.mdc

## Usage Examples
```bash
# Scope up a single task with regular strength
task-master scope-up --id=5

# Scope down multiple tasks with light strength
task-master scope-down --id=10,11,12 --strength=light

# Custom scoping with specific instructions
task-master scope-up --id=7 --prompt="Focus on adding more error handling and edge cases"
```

## Summary by CodeRabbit

* **New Features**
  * Introduced "Scope Up" and "Scope Down" actions in the UI and CLI, allowing users to dynamically increase or decrease task complexity with adjustable strength and optional custom prompts.
  * Added support for batch complexity adjustment on multiple tasks simultaneously.
  * Provided new documentation and usage examples for complexity adjustment commands.
  * Integrated complexity adjustment tools into MCP environments and the extension interface.

* **Bug Fixes**
  * Improved task template importing and tag preservation.

* **Documentation**
  * Added comprehensive guides on testing workflows and complexity adjustment features.

* **Tests**
  * Added unit tests for task complexity adjustment logic.

### CODE REVIEW

File: .changeset/floppy-news-buy.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+task-master scope-up --id=7 --prompt="Add more error handling and edge cases"
+```
+
+**Why use this?**
+- **Iterative refinement**: Adjust task complexity as your understanding evolves
+- **Project phase adaptation**: Scale tasks up for implementation, down for planning
+- **Team coordination**: Adjust complexity based on team member experience levels
+- **Milestone alignment**: Fine-tune tasks to match project phase requirements
+
+Perfect for agile workflows where task requirements change as you learn more about the problem space.
```
[the remark this answers is not in the document]
Comment (eyaltoledano) [reply .changeset/floppy-news-buy.md:38]: what the fuck is this garbage? I wrote this the way I did on purpose

File: .changeset/fix-tag-complexity-detection.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+"task-master-ai": patch
+---
+
```
[the remark this answers is not in the document]
Comment (Crunchyman-ralph) [reply .changeset/fix-tag-complexity-detection.md]: I don't mind having it, lets ignore comments like that in the future.

File: mcp-server/src/tools/scope-down.js
Comment (Crunchyman-ralph) [inline mcp-server/src/tools/scope-down.js:1]: Feels weird to have 2 tools, one to scope up, and one to scope down. Feels like it should be one tool "Scope" and it can be of type "up" or "down" in the parameters...

File: scripts/modules/commands.js [hunk 3]
```diff
@@ -1386,6 +1389,258 @@ function registerCommands(programInstance) {
...
+		.option(
+			'-f, --file <file>',
+			'Path to the tasks file',
+			TASKMASTER_TASKS_FILE
+		)
+		.option(
+			'-i, --id <ids>',
+			'Comma-separated task/subtask IDs to scope up (required)'
+		)
+		.option(
```
Comment (Crunchyman-ralph) [inline scripts/modules/commands.js:1405]: same for CLI. I think it should be scope --up or --down simply

File: apps/extension/src/webview/hooks/useTaskQueries.ts [hunk 4]
```diff
@@ -227,3 +227,119 @@ export function useUpdateSubtask() {
...
+			console.log(
+				'🔄 All task queries invalidated for scoped up task:',
+				variables.taskId
+			);
+		}
+	});
+}
+
+// Hook to scope down task complexity
+export function useScopeDownTask() {
```
Comment (Crunchyman-ralph) [inline apps/extension/src/webview/hooks/useTaskQueries.ts:290]: same here for the hook, how easy would it be if it was just one hook for the scope mutation.

File: apps/extension/src/webview/hooks/useTaskQueries.ts [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+				queryKey: taskKeys.all
+			});
+
+			console.log(
+				'🔄 All task queries invalidated for scoped down task:',
+				variables.taskId
+			);
+		}
+	});
+}
```
[the remark this answers is not in the document]
Comment (Crunchyman-ralph) [reply apps/extension/src/webview/hooks/useTaskQueries.ts:345]: Nice suggestion!

File: scripts/modules/task-manager/scope-adjustment.js [hunk 6]
```diff
@@ -0,0 +1,828 @@
...
+ * These represent work that has been started or intentionally set by the user
+ */
+const PRESERVE_STATUSES = [
+	'done',
+	'in-progress',
+	'review',
+	'cancelled',
+	'deferred',
+	'blocked'
+];
```
Comment (Crunchyman-ralph) [inline scripts/modules/task-manager/scope-adjustment.js:39]: I believe we have the statuses defined somewhere in src/constants, use those instead.

File: scripts/modules/task-manager/scope-adjustment.js [hunk 7]
```diff
@@ -0,0 +1,828 @@
...
+	generateTextService
+} from '../ai-services-unified.js';
+import { findTaskById, taskExists } from '../task-manager.js';
+import analyzeTaskComplexity from './analyze-task-complexity.js';
+import { findComplexityReportPath } from '../../../src/utils/path-utils.js';
+
+/**
+ * Valid strength levels for scope adjustments
+ */
+const VALID_STRENGTHS = ['light', 'regular', 'heavy'];
```
Comment (Crunchyman-ralph) [inline scripts/modules/task-manager/scope-adjustment.js:26]: should probably move that to src/constants ? but if we're not going to use it anywhere other than here, then we keep it here, though I saw it being used in mcp, cli, etc. so it would be smart to put it in src/constants

File: scripts/modules/task-manager/scope-adjustment.js [hunk 8]
```diff
@@ -0,0 +1,828 @@
...
+	'cancelled',
+	'deferred',
+	'blocked'
+];
+
+/**
+ * Statuses that should be regenerated during subtask regeneration
+ * These represent work that hasn't been started yet
+ */
+const REGENERATE_STATUSES = ['pending'];
```
Comment (Crunchyman-ralph) [inline scripts/modules/task-manager/scope-adjustment.js:45]: we have enums in src/constants for statuses, this is good to use here.

File: scripts/modules/task-manager/scope-adjustment.js [hunk 9]
```diff
@@ -0,0 +1,828 @@
...
+		if (!complexityReportPath) {
+			log('warn', 'No complexity report found - cannot re-analyze complexity');
+			return null;
+		}
+
+		// Use analyze-task-complexity to re-analyze just this task
+		const analysisOptions = {
+			file: tasksPath,
+			output: complexityReportPath,
+			id: task.id.toString(), // Analyze only this specific task
```
[the remark this answers is not in the document]
Comment (Crunchyman-ralph) [reply scripts/modules/task-manager/scope-adjustment.js:89]: nice suggestion, I know we've had bugs like these in the past!

File: scripts/modules/task-manager/scope-adjustment.js [hunk 10]
```diff
@@ -0,0 +1,828 @@
...
+	const isUp = direction === 'up';
+	const strengthDescriptions = {
+		light: isUp ? 'minor enhancements' : 'slight simplifications',
+		regular: isUp
+			? 'moderate complexity increases'
+			: 'moderate simplifications',
+		heavy: isUp ? 'significant complexity additions' : 'major simplifications'
+	};
+
+	let basePrompt = `You are tasked with adjusting the complexity of a task.
```
Comment (Crunchyman-ralph) [inline scripts/modules/task-manager/scope-adjustment.js:421]: we have a place where we put prompts in src/prompts, would be nice to keep that structure, will also allow you to pollute this file less. 800 lines in a file is a lot.

File: scripts/modules/task-manager/scope-adjustment.js [hunk 11]
```diff
@@ -0,0 +1,828 @@
...
+
+	if (outputFormat === 'text') {
+		log('info', `Successfully scoped up ${updatedTasks.length} task(s)`);
+	}
+
+	return {
+		updatedTasks,
+		telemetryData: combinedTelemetryData
+	};
+}
```
[the remark this answers is not in the document]
Comment (Crunchyman-ralph) [reply scripts/modules/task-manager/scope-adjustment.js:678]: I agree! great suggestion once again !

**Review by Crunchyman-ralph — CHANGES_REQUESTED** [review]
Really great feature, but I see a lot of duplication patterns between scope up and scope down.

Lets quickly refactor this part to reduce the size of the code, otherwise everything looks really good and I love how you used the different parts of our already existing functions to create something as needed as this scoping feature!

I have also taken a look at coderabbit's comments and have kept the ones I feel are relevant!
